### PR TITLE
Fed Avg

### PR DESCRIPTION
**Pull request by ZidongLiu**
- [x] basic fed avg implementation
- [x] isort
- [x] ruff format
- [x] test with small models
- [x] test with large models
- [x] test with 1/3/5 clients
- [x] check if momentum != 0 works

### CODE REVIEW

**Comment by ZidongLiu** [comment]
works for mnist
image

cifar10 can train to 40%
image

125m on sst2 can be trained to 85%
image

**Comment by RogerRogerUSC** [comment]
Can we use the name of the main file as "fo_fl_main.py"? Because we may add other first-order federated learning algorithms in the future.

**Comment by ZidongLiu** [comment]
> Can we use the name of the main file as "fo_fl_main.py"? Because we may add other first-order federated learning algorithms in the future.

@RogerRogerUSC just changed

File: fed_avg/server.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        for index in sampled_client_indices:
+            client = self.clients[index]
+            client.pull_model(self.server_model)
+            client_loss, client_accuracy = client.local_update(self.local_update_steps)
+            step_train_loss.update(client_loss)
+            step_train_accuracy.update(client_accuracy)
+            for i, p in enumerate(client.model.parameters()):
+                running_sum[i] += p.to(self.device)
+
+        self.update_server_model(running_sum)
```
Comment (BichengYing) [inline fed_avg/server.py:73]: I would suggest to make this out. Make the function like this
```python
self.aggregate_client_models([self.client[index] for index in sampled_client_indices])
```
Benefits are
1. This will make the future RPC or parallelism refactor cleaner 
2. Move the `running_sum` into sub-rountine so that the existence time is shorter
3. Cleaner structure and more flexible for future other federated learning extension.

File: fo_fl_main.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            [
+                get_args_str(args),
+                server.server_model.model_name,
+                model_helpers.get_current_datetime_str(),
+            ]
+        )
+        writer = SummaryWriter(
+            path.join(
+                "tensorboards",
+                "cezo_fl",
```
Comment (BichengYing) [inline fo_fl_main.py:85]: Should we change this to `fed_avg`?

File: fo_fl_main.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+            if args.adjust_perturb:
+                if ite == 500:
+                    server.set_learning_rate(args.lr * 0.8)
+                    server.set_perturbation(args.num_pert * 2)
+                elif ite == 1000:
+                    server.set_learning_rate(args.lr * 0.5)
+                    server.set_perturbation(args.num_pert * 4)
+                elif ite == 2000:
+                    server.set_learning_rate(args.lr * 0.3)
+                    server.set_perturbation(args.num_pert * 8)
```
Comment (BichengYing) [inline fo_fl_main.py:106]: delete this

File: fo_fl_main.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
+from os import path
+
+import torch
+from tensorboardX import SummaryWriter
+from tqdm import tqdm
+
+from cezo_fl.fl_helpers import get_client_name
+from cezo_fl.util import model_helpers
+from config import get_args_str, get_params
+from decomfl_main import prepare_settings_underseed
```
Comment (BichengYing) [inline fo_fl_main.py:10]: avoid import other `main`. Move this to `cezo_fl.util`?

**Comment by ZidongLiu** [comment]
still works
image

File: fed_avg/server.py (hunk 1 above)
Comment (ZidongLiu) [reply fed_avg/server.py:73]: @BichengYing 
Let me know if we want to do the following:

the suggested will have a n_client * model_size memory usage.
running sum will be 2 * model_size.

Comment (BichengYing) [reply fed_avg/server.py:73]: I don't understand. that should not change the memory consumption at all but just the coding refactors

**Comment by ZidongLiu** [comment]
fo_fl_main.py still works
image

decom_fl_main.py
image

**Comment by ZidongLiu** [comment]
@BichengYing I want to merge this PR, I noticed you merged a `run_experiments.sh`, I will leave it to you to decide if we need to keep that file. 

You can merge the PR if everything else looks good.

File: cezo_fl/util/model_helpers.py [hunk 5]
```diff
@@ -73,3 +77,14 @@ def get_trainable_model_parameters(
     for param in model.parameters():
         if param.requires_grad:
             yield param
+
+
+def model_forward(
```
Comment (BichengYing) [inline cezo_fl/util/model_helpers.py:82]: This is not sufficient for the generation model now unfortunately. We need a class or something equivalently to hold the generation parameters.

File: cezo_fl/util/prepare_settings.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+from cezo_fl.models.cnn_fashion import CNN_FMNIST
+from cezo_fl.models.cnn_mnist import CNN_MNIST
+from cezo_fl.models.lenet import LeNet
+from cezo_fl.models.lstm import CharLSTM
+from cezo_fl.random_gradient_estimator import RandomGradientEstimator as RGE
+from cezo_fl.util.language_utils import LM_TEMPLATE_MAP, SUPPORTED_LLM, get_lm_loss
+from cezo_fl.util.metrics import accuracy
+
+
+def prepare_settings_underseed(args, device, server_or_client: str = "server"):
```
Comment (BichengYing) [inline cezo_fl/util/prepare_settings.py:18]: I didn't move this function into settings because of this args, which is unstructured and can be anything. I would suggest to add another intermedia buffer between this function and args so that we don't need to worry about this function when refactor `args`. Wdyt? (not necessary in this PR)

please add more type annotation and comment after moving into modules.

**Review by BichengYing — CHANGES_REQUESTED** [review]
yeah, as long as the generation mode issue is resolved, let's check-in it.

About that script, it is uploaded by @RogerRogerUSC. I am ok either way. (The filename is too generic though)
FYI, the FedAvg cannot run the generation dataset. Is that okay? @RogerRogerUSC

File: cezo_fl/util/model_helpers.py (hunk 5 above)
Comment (RogerRogerUSC) [reply cezo_fl/util/model_helpers.py:82]: @BichengYing, it is better to support generation tasks in FedAvg.

File: cezo_fl/util/prepare_settings.py (hunk 6 above)
Comment (ZidongLiu) [reply cezo_fl/util/prepare_settings.py:18]: https://github.com/ZidongLiu/DeComFL/issues/64, created a issue.
I will make a dataclass for args in another PR

Comment (ZidongLiu) [reply cezo_fl/util/prepare_settings.py:18]: which can be used for typing all args

**Comment by ZidongLiu** [comment]
@BichengYing take a look to see if you like the new design

**Comment by BichengYing** [comment]
I think this structure looks good. Two comments:
- Remember to make sure the model inference is compatible for the llm generation mode.
- Prepare the settings becomes bigger and bigger. In the future, we can consider to split this into more smaller pieces and remove the dependency on the `args` directly.

**Comment by ZidongLiu** [comment]
experienced an unexpected different training process vs [30f6f5b](https://github.com/ZidongLiu/DeComFL/pull/60/commits/30f6f5b5fc7e7b0f4572917f71e9d053f01867e6)
need to investigate

python fo_fl_main.py --dataset=mnist --lr=1e-3
image

File: cezo_fl/util/prepare_settings.py [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+
+def get_torch_dtype(args_model_dtype: Literal["float32", "float16", "bfloat16"]) -> torch.dtype:
+    return {
+        "float32": torch.float32,
+        "float16": torch.float16,
+        "bfloat16": torch.bfloat16,
+    }[args_model_dtype]
+
+
+def get_random_gradient_estimator(args, model, device):
```
Comment (BichengYing) [inline cezo_fl/util/prepare_settings.py:36]: nit: type annotation

File: cezo_fl/util/prepare_settings.py [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+def get_model_and_optimizer(
+    dataset: str,
+    model_dtype: Literal["float32", "float16", "bfloat16"],
+    lr: float,
+    momentum: float,
+    large_model: str | None,
+    lora: bool = False,
+    lora_alpha: int = 16,
+    lora_r: int = 8,
+):
```
Comment (BichengYing) [inline cezo_fl/util/prepare_settings.py:62]: nit: return type annotation

File: cezo_fl/util/prepare_settings.py [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+    if dataset == "mnist":
+        model = CNN_MNIST().to(torch_dtype)
+        optimizer = torch.optim.SGD(
+            model_helpers.get_trainable_model_parameters(model),
+            lr=lr,
+            weight_decay=1e-5,
+            momentum=momentum,
+        )
+
+        # scheduler = torch.optim.lr_scheduler.ExponentialLR(optimizer, gamma=0.8)
```
Comment (BichengYing) [inline cezo_fl/util/prepare_settings.py:74]: let's delete this. If we need scheduler, we can make another function dedicated for scheduler.

File: cezo_fl/util/prepare_settings.py [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+                lr=lr,
+                momentum=0,
+                weight_decay=0,
+            )
+        else:
+            raise ValueError(f"Dataset {dataset} is not supported")
+    else:
+        raise Exception(f"Dataset {dataset} is not supported")
+
+    return (
```
Comment (BichengYing) [inline cezo_fl/util/prepare_settings.py:142]: nit: `return model, optimizer`

File: cezo_fl/util/prepare_settings.py [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+            mu=args.mu,
+            num_pert=args.num_pert,
+            grad_estimate_method=args.grad_estimate_method,
+            device=device,
+            torch_dtype=get_torch_dtype(args.model_dtype),
+            # To save memory consumption, we have to use parameter-wise perturb + no_optim together.
+            sgd_only_no_optim=args.no_optim,
+            paramwise_perturb=args.no_optim,
+        )
+    else:
```
Comment (BichengYing) [inline cezo_fl/util/prepare_settings.py:49]: you don't need this here since you already have assert in the RGE.

File: cezo_fl/util/prepare_settings.py (hunk 7 above)
Comment (BichengYing) [reply cezo_fl/util/prepare_settings.py:36]: Optional: this is also a good start point to decouple it from `args`:

```python
@dataclass
class RGE_Params:
     mu: float
     num_pert: int
     grad_estimate_method: str
     model_dtype: torch.dtype
     sgd_only_no_optim: bool
     paramwise_perturb: bool

     def parse_from_args():
          ....

def get_random_gradient_estimator(rge_param: RGE_Params, model, device):
        ....
```
This RGE_Params dataclass can be used in the RGE as well to avoid misalignment

**Comment by ZidongLiu** [comment]
> experienced an unexpected different training process vs [30f6f5b](https://github.com/ZidongLiu/DeComFL/pull/60/commits/30f6f5b5fc7e7b0f4572917f71e9d053f01867e6) need to investigate
> 
> python fo_fl_main.py --dataset=mnist --lr=1e-3 image

default parameter changes, using the following parameter gives the same training history
python fo_fl_main.py --dataset=mnist --lr=1e-3 --num-clients=5 --num-sample-clients=3 --local-update-steps=1 --train-batch-size=256 --test-batch-size=1000 --mu=0.0001 --momentum=0.9

image

File: cezo_fl/util/prepare_settings.py (hunk 7 above)
Comment (ZidongLiu) [reply cezo_fl/util/prepare_settings.py:36]: > Optional: this is also a good start point to decouple it from `args`:
> 
> ```python
> @dataclass
> class RGE_Params:
>      mu: float
>      num_pert: int
>      grad_estimate_method: str
>      model_dtype: torch.dtype
>      sgd_only_no_optim: bool
>      paramwise_perturb: bool
> 
>      def parse_from_args():
>           ....
> 
> def get_random_gradient_estimator(rge_param: RGE_Params, model, device):
>         ....
> ```
> 
> This RGE_Params dataclass can be used in the RGE as well to avoid misalignment

this is great suggestion! I will implement this change for all cli parameter groups in a later PR.

**Comment by ZidongLiu** [comment]
fo_fl_main has the same training history

image

File: cezo_fl/util/model_helpers.py [hunk 12]
```diff
@@ -73,3 +80,30 @@ def get_trainable_model_parameters(
...
+        assert isinstance(batch_inputs, LLMBatchInput)
+        return model(input_ids=batch_inputs.input_ids, attention_mask=batch_inputs.attention_mask)
+    elif isinstance(model, torch.nn.Module):
+        assert isinstance(batch_inputs, torch.Tensor)
+        return model(batch_inputs)
+    else:
+        raise Exception("This model type is not supported")
+
+
+def model_generate(model: LanguageModel, batch_inputs: LLMBatchInput, generation_kwargs: dict = {}):
```
Comment (BichengYing) [inline cezo_fl/util/model_helpers.py:96]: you should not use `generation_kwargs: dict = {}`, which is a common mistake in python. Even though I think the code here behaves correct, it is still dangerous.

it should be `generation_kwargs: dict | None = None` and inside the function we have `if generation_kwargs is None: generation_kwargs={}`. See the reason here: https://stackoverflow.com/questions/26320899/why-is-the-empty-dictionary-a-dangerous-default-value-in-python

File: cezo_fl/util/prepare_settings.py [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+    train_inference: Callable
+    test_inference: Callable
+
+
+@dataclass
+class MetricPacks:
+    train_loss: Callable
+    train_acc: Callable
+    test_loss: Callable
+    test_acc: Callable
```
Comment (BichengYing) [inline cezo_fl/util/prepare_settings.py:127]: looks good! Optional comment, we can let `train_acc` and `test_loss` to be optional and if the downstream code find it is None, just append None to the records?

**Review by BichengYing — APPROVED** [review]
LGTM!

File: cezo_fl/util/prepare_settings.py (hunk 13 above)
Comment (ZidongLiu) [reply cezo_fl/util/prepare_settings.py:127]: would prefer to keep this a strict structure, but we can make the train_acc and test_loss Callable do nothing when we define them and return Torch.Tensor(0.0)?

Like this

```
MetricPacks(train_loss = cross_entropy, lambda x, y: return torch.Tensor(0.0), test_loss =  lambda x, y: return torch.Tensor(0.0), test_acc = accuracy)
```

Comment (ZidongLiu) [reply cezo_fl/util/prepare_settings.py:127]: I see what you mean, you want to let downstream code to skip those steps

Comment (BichengYing) [reply cezo_fl/util/prepare_settings.py:127]: Yeah, I meant
```python
@dataclass
class MetricPacks:
    train_loss: Callable
    train_acc: Callable | None
    test_loss: Callable | None
    test_acc: Callable
```
But i think `test_loss =  lambda x, y: return torch.Tensor(0.0)` is also good.

File: cezo_fl/util/model_helpers.py (hunk 12 above)
Comment (ZidongLiu) [reply cezo_fl/util/model_helpers.py]: Thanks for pointing this out, resolved this issue by removing the default.
In our project so far, this is an required field for all cases.

File: cezo_fl/util/prepare_settings.py (hunk 13 above)
Comment (ZidongLiu) [reply cezo_fl/util/prepare_settings.py:127]: tried to implement a check for train_acc and test_loss in client and server, but found out we actually need these 2 fields in the tensorboard.
Thus decide to use `test_loss =  lambda x, y: return torch.Tensor(0.0)`
